File: domain/structure/chapter_element.py
"""
章节元素关联领域模型
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Optional


class ElementType(str, Enum):
    """元素类型"""
    CHARACTER = "character"
    LOCATION = "location"
    ITEM = "item"
    ORGANIZATION = "organization"
    EVENT = "event"


class RelationType(str, Enum):
    """关联类型"""
    APPEARS = "appears"      # 出场
    MENTIONED = "mentioned"  # 提及
    SCENE = "scene"          # 场景
    USES = "uses"            # 使用（道具）
    INVOLVED = "involved"    # 涉及（组织）
    OCCURS = "occurs"        # 发生（事件）


class Importance(str, Enum):
    """重要性"""
    MAJOR = "major"    # 主要
    NORMAL = "normal"  # 普通
    MINOR = "minor"    # 次要


@dataclass
class ChapterElement:
    """章节元素关联"""
    id: str
    chapter_id: str
    element_type: ElementType
    element_id: str
    relation_type: RelationType
    importance: Importance = Importance.NORMAL
    appearance_order: Optional[int] = None
    notes: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.now)

    def __post_init__(self):
        """类型转换"""
        if isinstance(self.element_type, str):
            self.element_type = ElementType(self.element_type)
        if isinstance(self.relation_type, str):
            self.relation_type = RelationType(self.relation_type)
        if isinstance(self.importance, str):
            self.importance = Importance(self.importance)

    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": self.id,
            "chapter_id": self.chapter_id,
            "element_type": self.element_type.value,
            "element_id": self.element_id,
            "relation_type": self.relation_type.value,
            "importance": self.importance.value,
            "appearance_order": self.appearance_order,
            "notes": self.notes,
            "created_at": self.created_at.isoformat(),
        }

    @classmethod
    def from_dict(cls, data: dict) -> "ChapterElement":
        """从字典创建"""
        return cls(
            id=data["id"],
            chapter_id=data["chapter_id"],
            element_type=ElementType(data["element_type"]),
            element_id=data["element_id"],
            relation_type=RelationType(data["relation_type"]),
            importance=Importance(data.get("importance", "normal")),
            appearance_order=data.get("appearance_order"),
            notes=data.get("notes"),
            created_at=datetime.fromisoformat(data["created_at"]) if isinstance(data.get("created_at"), str) else data.get("created_at", datetime.now()),
        )

File: frontend/src/api/chapterElement.ts
/**
 * 章节元素关联 API
 * 管理章节与 Bible 元素（人物/地点/道具/事件）的关联
 */

import { apiClient } from './config'

// ==================== 类型 ====================

export type ElementType = 'character' | 'location' | 'item' | 'organization' | 'event'
export type RelationType = 'appears' | 'mentioned' | 'scene' | 'uses' | 'involved' | 'occurs'
export type Importance = 'major' | 'normal' | 'minor'

export interface ChapterElementDTO {
  id: string
  chapter_id: string
  element_type: ElementType
  element_id: string
  relation_type: RelationType
  importance: Importance
  appearance_order: number | null
  notes: string | null
  created_at: string
}

export interface ChapterElementCreate {
  element_type: ElementType
  element_id: string
  relation_type: RelationType
  importance?: Importance
  appearance_order?: number
  notes?: string
}

// ==================== API ====================

export const chapterElementApi = {
  /** GET /api/v1/chapters/{chapter_id}/elements */
  getElements(chapterId: string, elementType?: ElementType): Promise<{ success: boolean; data: ChapterElementDTO[] }> {
    return apiClient.get(
      `/chapters/${chapterId}/elements`,
      { params: elementType ? { element_type: elementType } : undefined }
    ) as unknown as Promise<{ success: boolean; data: ChapterElementDTO[] }>
  },

  /** POST /api/v1/chapters/{chapter_id}/elements */
  addElement(chapterId: string, data: ChapterElementCreate): Promise<{ success: boolean; data: ChapterElementDTO }> {
    return apiClient.post(
      `/chapters/${chapterId}/elements`,
      data
    ) as unknown as Promise<{ success: boolean; data: ChapterElementDTO }>
  },

  /** PUT /api/v1/chapters/{chapter_id}/elements（批量替换） */
  batchUpdate(chapterId: string, elements: ChapterElementCreate[]): Promise<{ success: boolean; data: { updated_count: number; elements: ChapterElementDTO[] } }> {
    return apiClient.put(
      `/chapters/${chapterId}/elements`,
      { elements }
    ) as unknown as Promise<{ success: boolean; data: { updated_count: number; elements: ChapterElementDTO[] } }>
  },

  /** DELETE /api/v1/chapters/{chapter_id}/elements/{element_id} */
  deleteElement(chapterId: string, elementId: string): Promise<{ success: boolean; message: string }> {
    return apiClient.delete(
      `/chapters/${chapterId}/elements/${elementId}`
    ) as unknown as Promise<{ success: boolean; message: string }>
  },

  /** GET /api/v1/chapters/elements/{element_type}/{element_id}/chapters — 反向查哪些章用了该元素 */
  getElementChapters(elementType: ElementType, elementId: string): Promise<{ success: boolean; data: { appearance_count: number; chapters: unknown[] } }> {
    return apiClient.get(
      `/chapters/elements/${elementType}/${elementId}/chapters`
    ) as unknown as Promise<{ success: boolean; data: { appearance_count: number; chapters: unknown[] } }>
  },
}
